### PR TITLE
New data sources: aws_route53_resolver_rule and aws_route53_resolver_rules

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/8508.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
- New Data Source:  aws_route53_resolver_rule
- New Data Source:  aws_route53_resolver_rules
```

Output from acceptance testing:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsRoute53ResolverRule_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverRule_ -timeout 120m
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_basic (41.30s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags (252.57s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	252.604s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsRoute53ResolverRules_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverRules_ -timeout 120m
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_basic (20.62s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId (257.43s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	257.496s
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Related:
* https://github.com/terraform-providers/terraform-provider-aws/pull/8628

**Comment by ewbankkit** [comment]
TODO:
* Add `aws_route53_resolver_rules` data source

**Comment by ewbankkit** [comment]
Removed _WIP_.
Ready for review.

**Comment by ewbankkit** [comment]
Rebased to fix merge conflict.

File: aws/data_source_aws_route53_resolver_rule.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
```
Comment (nywilken) [inline aws/data_source_aws_route53_resolver_rule.go]: Lets fix the ordering of the github imports. I believe if you remove the blank line between the github imports gofmt should fix the ordering automatically.

File: aws/data_source_aws_route53_resolver_rules_test.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t); testAccPreCheckAWSRoute53Resolver(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceAwsRoute53ResolverRules_resolverEndpointId(rName1, rName2),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(ds1ResourceName, "resolver_rule_ids.#", "1"),
+					resource.TestCheckResourceAttr(ds2ResourceName, "resolver_rule_ids.#", "1"),
+					resource.TestCheckResourceAttr(ds3ResourceName, "resolver_rule_ids.#", "2"),
```
Comment (nywilken) [inline aws/data_source_aws_route53_resolver_rules_test.go]: Shared test cofiguration across data sources may be causing an issue when running all test in parallel. 

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverRule -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_basic (21.14s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_basic (45.18s)
--- FAIL: TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId (244.39s)
    testing.go:568: Step 0 error: Check failed: Check 3/4 error: data.aws_route53_resolver_rules.by_share_status: Attribute 'resolver_rule_ids.#' expected "2", got "3"
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags (256.00s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       256.053s
GNUmakefile:20: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

File: aws/data_source_aws_route53_resolver_rules.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/schema"
```
Comment (nywilken) [inline aws/data_source_aws_route53_resolver_rules.go]: Same issue here with go importing ordering

File: aws/data_source_aws_route53_resolver_rules.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		Read: dataSourceAwsRoute53ResolverRulesRead,
+
+		Schema: map[string]*schema.Schema{
+			"owner_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: validation.Any(
+					validateAwsAccountId,
+					// The owner of the default Internet Resolver rule.
+					validation.StringInSlice([]string{"Route 53 Resolver"}, false),
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply aws/data_source_aws_route53_resolver_rules.go]: Correct.
The rule that belongs to _Route 53 Resolver_ is always present and can't be deleted (and is handy for testing 😄).

File: aws/data_source_aws_route53_resolver_rules_test.go (hunk 2 above)
Comment (ewbankkit) [reply aws/data_source_aws_route53_resolver_rules_test.go]: Yes and also the existing state of any Route 53 Resolver Rules in the account.
I'll change the tests to be isolated from each other and any existing rules.

**Comment by ewbankkit** [comment]
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsRoute53ResolverRule_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverRule_ -timeout 120m
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_basic (41.96s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags (274.83s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	274.855s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsRoute53ResolverRules_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverRules_ -timeout 120m
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_basic (16.98s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId (428.15s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	428.186s
```

File: aws/data_source_aws_route53_resolver_rules.go (hunk 3 above)
Comment (ewbankkit) [reply aws/data_source_aws_route53_resolver_rules.go]: Done.

File: aws/data_source_aws_route53_resolver_rules_test.go (hunk 2 above)
Comment (ewbankkit) [reply aws/data_source_aws_route53_resolver_rules_test.go]: Done.

File: aws/data_source_aws_route53_resolver_rule.go (hunk 1 above)
Comment (ewbankkit) [reply aws/data_source_aws_route53_resolver_rule.go]: Done.

**Comment by ewbankkit** [comment]
@nywilken All review comments addressed.

**Comment by nywilken** [comment]
hi @ewbankkit sorry for the delayed turn around on this PR. I was still running into issues when running the tests for both data sources in parallel.
```
make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsRoute53ResolverRule'
```

I went ahead and fixed the test by changing the route53 resolver rule type in one of the test configurations to get it passing. I will merge this in and we can look to revisit the test if you think it no longer reflects your initial intent. 

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverRule -timeout 120m
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== PAUSE TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_basic
=== RUN   TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== PAUSE TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId
=== CONT  TestAccDataSourceAwsRoute53ResolverRules_basic
=== CONT  TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_basic (18.07s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_basic (42.41s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRule_ResolverEndpointIdWithTags (220.82s)
--- PASS: TestAccDataSourceAwsRoute53ResolverRules_ResolverEndpointId (220.96s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       221.002s
```

**Comment by nywilken** [comment]
Merged manually with additional commits.

**Pull request closed without merging**